gitlab/resource_gitlab_project_badge.go

```diff
@@ -0,0 +1,118 @@
+package gitlab
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+func resourceGitlabProjectBadge() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGitlabProjectBadgeCreate,
+		Read:   resourceGitlabProjectBadgeRead,
+		Update: resourceGitlabProjectBadgeUpdate,
+		Delete: resourceGitlabProjectBadgeDelete,
+
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"link_url": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"image_url": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"rendered_link_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"rendered_image_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceGitlabProjectBadgeCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	options := &gitlab.AddProjectBadgeOptions{
+		LinkURL:  gitlab.String(d.Get("link_url").(string)),
+		ImageURL: gitlab.String(d.Get("image_url").(string)),
+	}
+
+	log.Printf("[DEBUG] create gitlab project badge %q / %q", *options.LinkURL, *options.ImageURL)
+
+	badge, _, err := client.ProjectBadges.AddProjectBadge(project, options)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%d", badge.ID))
```

Comment:
The upstream API for badges requires a project ID as well as the badge ID. https://docs.gitlab.com/ee/api/project_badges.html#get-a-badge-of-a-project I think you need both of these IDs in your resource ID in order to avoid colliding IDs between projects.

Reply:
@armsnyder You are right. I will change it

Reply:
@armsnyder The resource will receive the project ID and every step will use it, so it won't collide.

Reply:
The same behavior of project hook: https://github.com/gitlabhq/terraform-provider-gitlab/blob/master/gitlab/resource_gitlab_project_hook.go#L133

Reply:
I'll explain the reason below, but my ask is:

Could you change the `gitlab_project_badge` resource's ID format to be `projectid:badgeid`, then add an importer, then modify the `resourceGitlabProjectBadgeRead` function to only reference the resource ID instead of by reading the `project` attribute? See `gitlab_project_mirror` for reference:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/8e2faa78634fe687a5d491e3370ca9fd206f2476/gitlab/resource_gitlab_project_mirror.go#L18-L20

https://github.com/gitlabhq/terraform-provider-gitlab/blob/8e2faa78634fe687a5d491e3370ca9fd206f2476/gitlab/resource_gitlab_project_mirror.go#L132-L137

### Explanation

Is there some GitLab documentation that says badge IDs are globally unique? Here's the scenario I want to avoid:
1. TF user creates `gitlab_project_badge.foo` for project `foo`. It gets an ID `1`.
2. TF user creates `gitlab_project_badge.bar` for project `bar`. It also gets an ID `1`.

Then there would be two of the same TF resource in the TF state with the same ID, which is not allowed by TF.

There is another reason for including the project ID in the resource ID, and that is it makes adding an importer a lot easier. **Ideally you want all of the IDs required to GET the upstream resource to be included in the TF resource ID**, so you can avoid needing to write a custom importer function, and use the `ImportStatePassthrough` helper instead: https://www.terraform.io/docs/extend/resources/import.html#importer-state-function

(We want all resources in this provider to have an importer.)

Reply:
@armsnyder Didn't think about how importer would work and it is a good enough reason. I will add the change tomorrow. Thank you for explaining

Reply:
@armsnyder Done. Could you review it again?